conans/model/graph_lock.py

```diff
@@ -2,297 +2,465 @@
+from conans import DEFAULT_REVISION_V1
+from conans.client.graph.graph import RECIPE_VIRTUAL, RECIPE_CONSUMER
 from conans.client.graph.python_requires import PyRequires
+from conans.client.graph.range_resolver import satisfying
 from conans.client.profile_loader import _load_profile
 from conans.errors import ConanException
 from conans.model.info import PACKAGE_ID_UNKNOWN
 from conans.model.options import OptionsValues
-from conans.model.ref import PackageReference, ConanFileReference
-from conans.model.version import Version
+from conans.model.ref import ConanFileReference
 from conans.util.files import load, save
 
 LOCKFILE = "conan.lock"
-LOCKFILE_VERSION = "0.3"
+LOCKFILE_VERSION = "0.4"
 
 
 class GraphLockFile(object):
 
     def __init__(self, profile_host, profile_build, graph_lock):
-        self.profile_host = profile_host
-        self.profile_build = profile_build
-        self.graph_lock = graph_lock
+        self._profile_host = profile_host
+        self._profile_build = profile_build
+        self._graph_lock = graph_lock
+
+    @property
+    def graph_lock(self):
+        return self._graph_lock
+
+    @property
+    def profile_host(self):
+        return self._profile_host
+
+    @property
+    def profile_build(self):
+        return self._profile_build
 
     @staticmethod
     def load(path, revisions_enabled):
         if not path:
             raise IOError("Invalid path")
         if not os.path.isfile(path):
-            p = os.path.join(path, LOCKFILE)
-            if not os.path.isfile(p):
-                raise ConanException("Missing lockfile in: %s" % path)
-            path = p
+            raise ConanException("Missing lockfile in: %s" % path)
         content = load(path)
         try:
-            return GraphLockFile.loads(content, revisions_enabled)
+            return GraphLockFile._loads(content, revisions_enabled)
         except Exception as e:
             raise ConanException("Error parsing lockfile '{}': {}".format(path, e))
 
+    def save(self, path):
+        serialized_graph_str = self._dumps(path)
+        save(path, serialized_graph_str)
+
     @staticmethod
-    def loads(text, revisions_enabled):
+    def _loads(text, revisions_enabled):
         graph_json = json.loads(text)
         version = graph_json.get("version")
         if version:
-            version = Version(version)
-            if version < "0.2":
-                raise ConanException("This lockfile was created with a previous incompatible "
+            if version != LOCKFILE_VERSION:
+                raise ConanException("This lockfile was created with an incompatible "
                                      "version. Please regenerate the lockfile")
             # Do something with it, migrate, raise...
-        profile_host = graph_json.get("profile_host") or graph_json["profile"]
+        profile_host = graph_json.get("profile_host", None)
         profile_build = graph_json.get("profile_build", None)
```

Comment:
I think this is not needed for this object

```suggestion
        profile_host = graph_json.get("profile_host")
        profile_build = graph_json.get("profile_build")
```